### PR TITLE
[2.2] Use non-pinnable buffer for zero byte read

### PR DESCRIPTION
**Pull request by benaadams**
Time is being spent Pinning 

image

and Unpinning

image

Kestrel uses prepinned buffers so shouldn't have these costs. It comes from pinning `Array.Empty<byte>()` in the zero-byte reads.

After using `Memory<byte>.Empty` these costs go away:

image

@sebastienros could you benchmark this; see if it makes a difference?

### CODE REVIEW

**Review by halter73 — APPROVED** [review]

**Comment by sebastienros** [comment]
Plaintext (pipelined) not impacted by this change.

Json goes back to the levels of 2.1, so actually reverting the negative impact of the ZBR PR.
DbFortunesEf is slightly better.

**Comment by benaadams** [comment]
Marked as [2.2] as it targeting that branch

**Comment by Eilon** [comment]
Moving to patch milestone.

**Comment by muratg** [comment]
@Eilon @natemcmaster I'm assuming it's fine to check-in to release/2.2 now right? (We're aiming to move Kestrel to mondorepo ASAP so it would be good to clear the PRs ASAP.

**Comment by natemcmaster** [comment]
If you merge this in the next few hours, it's fine to go into release/2.2. But you're racing against https://github.com/aspnet/AspNetCore/pull/4026. Once https://github.com/aspnet/AspNetCore/pull/4026 is merged, I'll be locking release/2.2 and master from any changes.

**Comment by halter73** [comment]
Alright. It's in.

**Comment by sebastienros** [comment]
I get it now, this time window only allowed for this PR not to be lost when you move to mondo-repo, and it will be part of the next servicing release for 2.2.

**Comment by Eilon** [comment]
This... wasn't approved for 2.2...

**Comment by Eilon** [comment]
@halter73 / @muratg - this change isn't approved for 2.2 or 2.2.1 yet. Can we please back it out until it's approved?

**Comment by muratg** [comment]
My apologies, this was my mistake. I guess I "imagined" that this PR was not to master branch.

Eilon is right. This is OK to check-in to master, but for release/2.2.1 we need to get shiproom approval. 

@halter73 could you revert it please? And we can talk offline to decide if we want to take this to shiproom.

**Comment by halter73** [comment]
Reverted.

**Comment by Eilon** [comment]
Thanks! We should keep an open PR though so that we can take it through shiproom. Unfortunately can't re-open a PR. Can we send a new one from the same fork or something?

**Comment by Tratcher** [comment]
We're moving to mondo-repo now, we'll need to create a new PR over there.

**Comment by benaadams** [comment]
mon**d**o-repo? Is that 'cos mono-repo would be confusing in the .NET world? 

Or am I just not up to date with mah lingo?

**Comment by Tratcher** [comment]
@benaadams see https://github.com/aspnet/AspNetCore/pull/4026

**Comment by benaadams** [comment]
> We're moving to mondo-repo now, we'll need to create a new PR over there.

In which branch? (master, release2.2 etc)

**Comment by Tratcher** [comment]
2.1 just moved, 2.2 is in flight, master is next.

**Comment by natemcmaster** [comment]
2.2 - https://github.com/aspnet/AspNetCore/pull/4031

**Comment by davidfowl** [comment]
@halter73 we should get this into master

**Comment by benaadams** [comment]
Zero byte read for AspNetCore master https://github.com/aspnet/AspNetCore/pull/4059

**Comment by vivmishra** [comment]
Approved for 2.2.1

**Pull request merged**
